pkg/task/meta.go

```diff
@@ -26,137 +26,210 @@ import (
-	// specified. Each task execution is based on a resource from this provided
-	// list of resources.
+}
+
+// toString returns a string representation of MetaTaskProps structure. In this
+// representation, each property is separated from its value via '='. In
+// addition each property=value pair is separated from other pair via '::'.
+//
+// Example:
+//  runNamespace=default::objectName=MySvc::retry=3,20s
+func (m MetaTaskProps) toString() string {
+	return fmt.Sprintf("runNamespace=%s::owner=%s::objectName=%s::options=%s::retry=%s",
+		m.RunNamespace,
+		m.Owner,
+		m.ObjectName,
+		m.Options,
+		m.Retry)
+}
+
+// selectOverride will override the current meta task properties from the given
+// if the given's properties has value
+func (m MetaTaskProps) selectOverride(given MetaTaskProps) MetaTaskProps {
+	namespace := strings.TrimSpace(given.RunNamespace)
+	if len(namespace) != 0 {
+		m.RunNamespace = namespace
+	}
+	owner := strings.TrimSpace(given.Owner)
+	if len(owner) != 0 {
+		m.Owner = owner
+	}
+	objectname := strings.TrimSpace(given.ObjectName)
+	if len(objectname) != 0 {
+		m.ObjectName = objectname
+	}
+	options := strings.TrimSpace(given.Options)
+	if len(options) != 0 {
+		m.Options = options
+	}
+	retry := strings.TrimSpace(given.Retry)
+	if len(retry) != 0 {
+		m.Retry = retry
+	}
+
+	return m
+}
+
+// MetaTaskSpec is the specifications of a MetaTask
+type MetaTaskSpec struct {
+	// MetaTaskIdentity provides the identity to this task
+	MetaTaskIdentity
+	// MetaTaskProps provides the task's meta related properties
+	MetaTaskProps
+	// Action representing this task
+	//
+	// e.g. get based task or list based task or put based task and so on
+	Action MetaTaskAction `json:"action"`
+	// RepeatWith sets one or more resources for repetitive execution.
+	// In other words a task template is executed multiple times based on each
+	// of the item present here.
 	RepeatWith RepeatWithResource `json:"repeatWith"`
 }
 
 type metaTaskExecutor struct {
-	// metaTask holds various metadata related info w.r.t the task
-	metaTask MetaTask
-	// identifier is a utility struct that enables a task's identity
-	// related operations
+	// metaTask holds the task's meta information
+	metaTask MetaTaskSpec
+	// identifier exposes a task's identity related operations
 	identifier taskIdentifier
 	// repeater exposes operations with respect to repetitive execution of this
 	// task
-	repeater repeatWithResourceExecutor
+	repeater repeatExecutor
+	// k8sClient will be used to make K8s API calls
+	k8sClient *m_k8s_client.K8sClient
 }
 
-// newMetaTaskExecutor provides a new instance of metaTaskExecutor
-func newMetaTaskExecutor(yml string, values map[string]interface{}) (*metaTaskExecutor, error) {
+// getMetaInstances is a utility function that provides required objects
+// to instantiate meta task executor
+func getMetaInstances(metaTaskYml string, values map[string]interface{}) (m MetaTaskSpec, i taskIdentifier, r repeatExecutor, err error) {
```

Comment:
This function returns too many values. Can we encapsulate it into a struct object or use some method like `setMetaInstances` to set the values in the `metaTaskExecutor` itself

Reply:
Yes. the code smells here. This will require a bit of refactoring.
It should deal more with single method interfaces & composition of these interfaces to start with.
Need to spend more time & effort on doing this. At the same time, all these, effective & simpler way to UT & IT and then cas template's usability related pain points can be redesigned into v1alpha2 or v1beta1 packages.

So IMO this should be built over iterations by taking into consideration multiple sprints.

Reply:
Sure.